State Machine Rewrite

cstyan opened the pull request:
It makes more sense for our implementation of ADB to be written as a state machine, as during various parts of the connection phase the protocol has to account for timeouts and return to the initial state. Unfortunately the daemon handling on the device/emulator is not smart enough to accept retransmits during the connection phase.

Here's a diagram: 
![State Machine Diagram](https://github.com/appium/node-adb-client/raw/fsm-rewrite/connectionStateMachine.png)

Once we're in the connected state we can open and run as many commands as we want using [ADB.runCommand](https://github.com/appium/node-adb-client/blob/fsm-rewrite/adb.js#L137). Still need to write tests for all of this.

cc: @jlipps @imurchie @scottdixon

imurchie commented on adb.js:
```diff
@@ -0,0 +1,162 @@
+    logExceptOnTest("Trying to find a usb device");
+    let adbDevices = [];
+    let usbDevices = usb.getDeviceList();
+    let deviceIndex = 0;
```
What is `deviceIndex` being used for? It seems to be incremented but never accessed.

imurchie commented on adb.js:
```diff
@@ -0,0 +1,162 @@
+
+  // search through a devices interfaces for an interface
+  // that can be used for ADB communications
+  static getAdbInterface (device) {
```
This method will return one of three values: `null`, `undefined`, or the adb interface. It would be better, for consistency's sake, to limit to two (either initializing `returnInterface` to `undefined`, or returning `null` at the end).

cstyan replied:
if there's ever a case where the `deviceDescriptor` or `configDescriptor` are null something is really wrong with the physical device, but I added a `return null` at the end of the function just to be safe

imurchie commented on adb.js:
```diff
@@ -0,0 +1,162 @@
+              this.state = SEND_PRIVATE_KEY;
+            }
+          } catch (e) {
+            if (e.errno) {
```
Is _any_ `errno` indicative of a timeout here? The other cases seem to have `errno === 2` to indicate a timeout.

imurchie commented on adb.js:
```diff
@@ -0,0 +1,162 @@
+    if (this.state === CONNECTED) {
+      await this.device.open(command);
+    }
+    // else log, not ready to run a command/open a stream
```
No indication back to the calling code that nothing has happened?

imurchie commented on gulpfile.js:
```diff
@@ -1,6 +1,11 @@
 var gulp = require('gulp'),
-    boilerplate = require('appium-gulp-plugins').boilerplate.use(gulp);
+  Q = require('q') ,
+  exec = Q.denodeify(require('child_process').exec),
```
Should probably keep to one Promise implementation.

imurchie commented:
This is pretty awesome! A few comments but overall I like the direction.

As an aside: you might find useful the `appium-support` package, which (in addition to other things) provides promisified versions of the generally used methods (like many of nodes `fs` methods). Then you don't need to deal with promisifying them yourself.

cstyan commented:
@imurchie Thanks for the review! I think I've addressed all your comments with the latest commits, I also opened a PR to `appium-support`.

What are your thoughts re: testing and TCP support?

cstyan commented:
Merging, going to start a new branch for tests

cstyan commented:
fixes issue #7
